### ISSUE TITLE
Question: Logits after tanh is taken even for last layer in RNN-P encoder

### ISSUE THREAD

**Issue by RSSharma246** [espnet/espnet#1304]
In https://github.com/espnet/espnet/blob/7f687347c472e3c7082d74456ebc82dc7126e15a/espnet/nets/pytorch_backend/rnn/encoders.py#L79,
I notice that the tanh activation is applied even to the final layer output in the RNNP encoder, and I was wondering if there is a reason why this is done ? Shouldn't the activation be on the pre-final layers only?

**Comment by sw005320** [comment]
Thanks for pointing out it. Frankly, I don't know which is better and it's worth considering to compare it. @simpleoier, could you use the an4/voxforge tasks and report the difference?

**Comment by simpleoier** [comment]
Sure.

**Comment by simpleoier** [comment]
Alright, I tested it. The improvement can be observed if no tanh activation is applied to the final layer of the RNNP encoder.
Here is the results for the current ESPnet (with tanh for the last encoder layer):
* hybrid CTC/attention using word-LM
```
grep -e Avg -e SPKR -m 2 exp/exp*/train_*0.5/decode*/result.txt                                                                                
exp/exp_version1/train_nodev_pytorch_train_mtlalpha0.5/decode_test_decode_ctcweight1.0_lm_word100/result.txt:|  SPKR     |  # Snt    # Wrd   |  Corr        Sub       Del       Ins       Err     S.Err   |
exp/exp_version1/train_nodev_pytorch_train_mtlalpha0.5/decode_test_decode_ctcweight1.0_lm_word100/result.txt:|  Sum/Avg  |   130      2565   |  88.0        5.0       7.1       0.7      12.7      60.0   |
exp/exp_version1/train_nodev_pytorch_train_mtlalpha0.5/decode_train_dev_decode_ctcweight1.0_lm_word100/result.txt:|   SPKR     |  # Snt     # Wrd   |   Corr     Sub        Del        Ins       Err      S.Err   |
exp/exp_version1/train_nodev_pytorch_train_mtlalpha0.5/decode_train_dev_decode_ctcweight1.0_lm_word100/result.txt:|   Sum/Avg  |   100       1915   |   81.3     7.2       11.5        0.9      19.6       75.0   |
```
* CTC only using word-LM
```
grep -e Avg -e SPKR -m 2 exp/exp*/train_*1.0/decode*/result.txt                                                                                
exp/exp_version1/train_nodev_pytorch_train_mtlalpha1.0/decode_test_decode_ctcweight1.0_lm_word100/result.txt:|  SPKR     |  # Snt    # Wrd   |  Corr        Sub       Del       Ins       Err     S.Err   |
exp/exp_version1/train_nodev_pytorch_train_mtlalpha1.0/decode_test_decode_ctcweight1.0_lm_word100/result.txt:|  Sum/Avg  |   130      2565   |  91.9        3.6       4.5       0.8       8.9      53.1   |
exp/exp_version1/train_nodev_pytorch_train_mtlalpha1.0/decode_train_dev_decode_ctcweight1.0_lm_word100/result.txt:|   SPKR     |  # Snt     # Wrd   |   Corr     Sub        Del        Ins       Err      S.Err   |
exp/exp_version1/train_nodev_pytorch_train_mtlalpha1.0/decode_train_dev_decode_ctcweight1.0_lm_word100/result.txt:|   Sum/Avg  |   100       1915   |   85.9     5.4        8.7        1.5      15.6       68.0   |
```
Here is the results for the current ESPnet (disabling tanh for the last encoder layer):
* hybrid CTC/attention using word-LM
```
grep -e Avg -e SPKR -m 2 exp/train_*/decode*/result.txt 
exp/train_nodev_pytorch_train_mtlalpha0.5/decode_test_decode_ctcweight1.0_lm_word100/result.txt:|  SPKR     |  # Snt    # Wrd   |  Corr        Sub       Del     Ins       Err     S.Err   |
exp/train_nodev_pytorch_train_mtlalpha0.5/decode_test_decode_ctcweight1.0_lm_word100/result.txt:|  Sum/Avg  |   130      2565   |  88.7        4.0       7.3     0.6      11.9      57.7   |
exp/train_nodev_pytorch_train_mtlalpha0.5/decode_train_dev_decode_ctcweight1.0_lm_word100/result.txt:|   SPKR     |  # Snt     # Wrd   |   Corr       Sub   Del        Ins       Err      S.Err   |
exp/train_nodev_pytorch_train_mtlalpha0.5/decode_train_dev_decode_ctcweight1.0_lm_word100/result.txt:|   Sum/Avg  |   100       1915   |   81.1       6.6  12.3        1.1      20.0       74.0   |
```
* CTC only using word-LM
```
grep -e Avg -e SPKR -m 2 exp/train_*1.0/decode*/result.txt 
exp/train_nodev_pytorch_train_mtlalpha1.0/decode_test_decode_ctcweight1.0_lm_word100/result.txt:|  SPKR     |  # Snt    # Wrd   |  Corr        Sub       Del     Ins       Err     S.Err   |
exp/train_nodev_pytorch_train_mtlalpha1.0/decode_test_decode_ctcweight1.0_lm_word100/result.txt:|  Sum/Avg  |   130      2565   |  92.8        3.0       4.2     1.1       8.3      50.0   |
exp/train_nodev_pytorch_train_mtlalpha1.0/decode_train_dev_decode_ctcweight1.0_lm_word100/result.txt:|   SPKR     |  # Snt     # Wrd   |   Corr       Sub   Del        Ins       Err      S.Err   |
exp/train_nodev_pytorch_train_mtlalpha1.0/decode_train_dev_decode_ctcweight1.0_lm_word100/result.txt:|   Sum/Avg  |   100       1915   |   87.3       4.5   8.2        1.5      14.3       59.0   |
```

**Comment by sw005320** [comment]
can you test a few more recpes? If we still observe it, we can just use this non tanh one.

**Comment by simpleoier** [comment]
No problem. I was going to.

**Comment by RSSharma246** [comment]
Do you have numbers for the attention part, i.e., decoding ctc_weight !=0 for the mtl=0.5 model?
Might be interesting to see if those numbers improve and by how much?
Thanks

**Comment by simpleoier** [comment]
> Do you have numbers for the attention part, i.e., decoding ctc_weight !=0 for the mtl=0.5 model?

Do you mean decoding ctc_weight!=1.0?

**Comment by RSSharma246** [comment]
Sorry, typo, that's what I meant

**Comment by simpleoier** [comment]
@sw005320 Here, I put the results on WSJ recipe. There are still improvements when disabling the tanh for the last encoder layer.
* Here is the results for the current ESPnet (with tanh for the last encoder layer):
```
grep -e Avg -e SPKR -m 2 exp/version1/train_*_rnn*/decode*/result.wrd.txt
exp/version1/train_si284_pytorch_train_rnn_no_preprocess/decode_test_dev93_decode_rnn_lm_word65000/result.wrd.txt:|   SPKR     |   # Snt     # Wrd   |   Corr       Sub         Del        Ins        Err      S.Err   |
exp/version1/train_si284_pytorch_train_rnn_no_preprocess/decode_test_dev93_decode_rnn_lm_word65000/result.wrd.txt:|   Sum/Avg  |    503       8234   |   91.6       7.2         1.2        1.3        9.6       61.8   |
exp/version1/train_si284_pytorch_train_rnn_no_preprocess/decode_test_eval92_decode_rnn_lm_word65000/result.wrd.txt:|   SPKR     |   # Snt     # Wrd    |   Corr        Sub        Del        Ins         Err      S.Err   |
exp/version1/train_si284_pytorch_train_rnn_no_preprocess/decode_test_eval92_decode_rnn_lm_word65000/result.wrd.txt:|   Sum/Avg  |    333       5643    |   95.2        4.3        0.6        1.1         5.9       43.5   |
```
* Here is the results for the current ESPnet (disabling tanh for the last encoder layer):
```
grep -e Avg -e SPKR -m 2 exp/train_*_rnn*/decode*/result.wrd.txt
exp/train_si284_pytorch_train_rnn_no_preprocess/decode_test_dev93_decode_rnn_lm_word65000/result.wrd.txt:|   SPKR     |   # Snt     # Wrd   |   Corr        Sub         Del        Ins        Err      S.Err   |
exp/train_si284_pytorch_train_rnn_no_preprocess/decode_test_dev93_decode_rnn_lm_word65000/result.wrd.txt:|   Sum/Avg  |    503       8234   |   92.5        6.5         1.0        1.2        8.7       59.0   |
exp/train_si284_pytorch_train_rnn_no_preprocess/decode_test_eval92_decode_rnn_lm_word65000/result.wrd.txt:|   SPKR     |   # Snt     # Wrd    |   Corr       Sub        Del        Ins         Err      S.Err   |
exp/train_si284_pytorch_train_rnn_no_preprocess/decode_test_eval92_decode_rnn_lm_word65000/result.wrd.txt:|   Sum/Avg  |    333       5643    |   95.5       3.9        0.6        0.9         5.4       44.7   |
```

**Comment by sw005320** [comment]
This is very significant.
Could you make a PR?
